File: src/code/python/graph/dijkstra.py
from heapq import heappop, heappush


def dijkstras(graph: list[list[tuple[int, int]]], source: int) -> list[int]:
    n = len(graph)
    distances = [float('inf')] * n
    distances[source] = 0
    heap = [(0, source)]

    while heap:
        curr_dist, node = heappop(heap)

        if curr_dist > distances[node]:
            continue

        for neighbor, weight in graph[node]:
            dist = curr_dist + weight

            if dist < distances[neighbor]:
                distances[neighbor] = dist
                heappush(heap, (dist, neighbor))

    return distances

File: src/code/ruby/graph/dijkstra.rb
require 'priority_queue'

def dijkstras(graph, source)
  n         = graph.size
  distances = Array.new(n, Float::INFINITY)

  distances[source] = 0

  heap = PriorityQueue.new
  heap[source] = 0

  until heap.empty?
    node, curr_dist = heap.delete_min
    next if curr_dist > distances[node]

    graph[node].each do |neighbor, weight|
      dist = curr_dist + weight
      if dist < distances[neighbor]
        distances[neighbor] = dist
        heap[neighbor] = dist
      end
    end
  end

  distances
end
